test/src/test/java/jenkins/model/GlobalBuildDiscarderTest.java

```diff
@@ -0,0 +1,109 @@
+            j.buildAndAssertSuccess(p);
+            j.buildAndAssertSuccess(p);
+            j.buildAndAssertSuccess(p);
+            j.buildAndAssertSuccess(p);
+            Assert.assertArrayEquals("all 5 builds exist", p.getBuilds().stream().mapToInt(Run::getNumber).toArray(), new int[]{5, 4, 3, 2, 1});
+        }
+
+        { // job build discarder
+            GlobalBuildDiscarderConfiguration.get().getConfiguredBuildDiscarders().add(new JobGlobalBuildDiscarderStrategy());
+            p.setBuildDiscarder(new LogRotator(null, "3", null, null));
+            Assert.assertArrayEquals("all 5 builds exist", p.getBuilds().stream().mapToInt(Run::getNumber).toArray(), new int[]{5, 4, 3, 2, 1});
+
+            ExtensionList.lookupSingleton(BackgroundGlobalBuildDiscarder.class).execute(TaskListener.NULL);
+            Assert.assertArrayEquals("only 3 builds left", p.getBuilds().stream().mapToInt(Run::getNumber).toArray(), new int[]{5, 4, 3});
+
+            j.buildAndAssertSuccess(p);
+            Assert.assertArrayEquals("still only 3 builds", p.getBuilds().stream().mapToInt(Run::getNumber).toArray(), new int[]{6, 5, 4});
+
+            p.setBuildDiscarder(null);
+            j.buildAndAssertSuccess(p);
+            j.buildAndAssertSuccess(p);
+            Assert.assertArrayEquals("5 builds again", p.getBuilds().stream().mapToInt(Run::getNumber).toArray(), new int[]{8, 7, 6, 5, 4});
+
+            ExtensionList.lookupSingleton(BackgroundGlobalBuildDiscarder.class).execute(TaskListener.NULL);
+            Assert.assertArrayEquals("still 5 builds", p.getBuilds().stream().mapToInt(Run::getNumber).toArray(), new int[]{8, 7, 6, 5, 4});
+        }
+
+        { // global build discarder
+            GlobalBuildDiscarderConfiguration.get().getConfiguredBuildDiscarders().add(new SimpleGlobalBuildDiscarderStrategy(new LogRotator(null, "2", null, null)));
+            ExtensionList.lookupSingleton(BackgroundGlobalBuildDiscarder.class).execute(TaskListener.NULL);
+            Assert.assertArrayEquals("newest 2 builds", p.getBuilds().stream().mapToInt(Run::getNumber).toArray(), new int[]{8, 7});
+            j.buildAndAssertSuccess(p);
+            j.buildAndAssertSuccess(p);
+
+            // run global discarders once a build finishes
+            Assert.assertArrayEquals("2 builds because of BackgroundBuildDiscarderListener", p.getBuilds().stream().mapToInt(Run::getNumber).toArray(), new int[]{10, 9});
+
+            GlobalBuildDiscarderConfiguration.get().getConfiguredBuildDiscarders().clear();
+            GlobalBuildDiscarderConfiguration.get().getConfiguredBuildDiscarders().add(new SimpleGlobalBuildDiscarderStrategy(new LogRotator(null, "1", null, null)));
+
+            // apply global config changes periodically
+            ExtensionList.lookupSingleton(BackgroundGlobalBuildDiscarder.class).execute(TaskListener.NULL);
+            Assert.assertArrayEquals("2 builds again", p.getBuilds().stream().mapToInt(Run::getNumber).toArray(), new int[]{10});
```

Comment:
Could save a few characters for the reader and make it look a bit nicer if you static imported Assert.assertArrayEquals throughout this file